### ISSUE TITLE
normal_transform_pixel: size == 1 gives a 2e14 scale and size == 0 silently mirrors - the executed root cause behind #3929's NaN crops

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#3957]
> **Scope note, please read first.** This issue does not propose changing the `2/(size − 1)` scaling or the corner-aligned convention. **#3904 remains reserved for its assigned contributor**; the defect here is the degenerate-denominator branch (`size == 1` / `size == 0` / negative), which is wrong under **every** `align_corners` setting. The NaN reproduced below is an `align_corners=True` execution, i.e. it is present even in the branch #3904 leaves alone, and fixing #3904 as described in its body would not touch the `size == 1` / `size == 0` code at all. No code change to the scaling formula is proposed here.

`normal_transform_pixel` guards its denominator with an epsilon substitution instead of a range check (`kornia/geometry/conversions.py:1318-1322`):

```python
# prevent divide by zero bugs
width_denom: float = eps if width == 1 else width - 1.0
height_denom: float = eps if height == 1 else height - 1.0

sx: float = 2.0 / width_denom
sy: float = 2.0 / height_denom
```

with `eps = 1e-14`. Three separate failures follow, all silent and all `isfinite`:

- **`size == 1`** → `2 / 1e-14 = 2e14`. Not an error, not a finite fallback — a scale factor fourteen orders of magnitude out, which downstream becomes NaN.
- **`size == 0`** → `2 / (0 − 1) = −2.0`, a **sign-flipped (mirroring)** normalization. This is the one case that is literally a divide-by-zero candidate, and it is the one case `eps` does *not* guard: `eps` is consulted only in the `== 1` branch. The docstring says `eps: epsilon to prevent divide-by-zero errors`.
- **negative sizes** → accepted silently (`width = −3` → `sx = −0.5`).

`normal_transform_pixel3d` has the same three branches with the same behavior, and `normalize_homography`, `denormalize_homography` and `normalize_homography3d` all call these helpers, so the degeneracy propagates through the public API.

## Repro

```python
import torch
from kornia.geometry.conversions import (
    denormalize_homography,
    normal_transform_pixel,
    normal_transform_pixel3d,
    normalize_homography,
    normalize_homography3d,
)
from kornia.geometry.transform import crop_by_transform_mat, warp_perspective

print("# 1. the helper itself")
print("normal_transform_pixel(4, 1) =\n", normal_transform_pixel(4, 1))
print("normal_transform_pixel(1, 1) flat =", normal_transform_pixel(1, 1).flatten().tolist())
for h, w in [(4, 0), (0, 5), (0, 0), (4, -3)]:
    M = normal_transform_pixel(h, w)
    print(f"  h={h},w={w} -> sx={M[0,0,0].item()}, sy={M[0,1,1].item()}, finite={bool(torch.isfinite(M).all())}")

print("\n# 2. eps guards only the ==1 branch, not the ==0 one")
print("  eps=1.0,  W==1 -> sx =", normal_transform_pixel(4, 1, eps=1.0)[0, 0, 0].item())
print("  eps=1.0,  W==0 -> sx =", normal_transform_pixel(4, 0, eps=1.0)[0, 0, 0].item())
print("  eps=1e-14,W==5 -> sx =", normal_transform_pixel(4, 5, eps=1e-14)[0, 0, 0].item())

print("\n# 3. it propagates into the public homography API")
I = torch.eye(3)[None]
print("normalize_homography(I,(4,1),(4,5))   =", normalize_homography(I, (4, 1), (4, 5)).flatten().tolist())
print("denormalize_homography(I,(4,1),(4,5)) =", denormalize_homography(I, (4, 1), (4, 5)).flatten().tolist())
print("normalize_homography(I,(4,0),(4,5))   =", normalize_homography(I, (4, 0), (4, 5)).flatten().tolist())

print("\n# 4. and into 3D, where depth==1 gives a numerically non-invertible matrix")
o = normalize_homography3d(torch.eye(4)[None], (1, 4, 5), (3, 8, 9))[0]
print("normalize_homography3d(I,(1,4,5),(3,8,9)) row2 =", o[2].tolist())
print("  o[2,2] =", o[2, 2].item(), " det =", torch.linalg.det(o).item())
print("  torch.linalg.inv succeeds:", bool(torch.isfinite(torch.linalg.inv(o)).all()))
print("normal_transform_pixel3d(1,4,5) diag =", [normal_transform_pixel3d(1, 4, 5)[0, i, i].item() for i in range(3)])

print("\n# 5. the user-visible end of the chain: NaN out of warp_perspective (align_corners=TRUE)")
x = torch.arange(25.0).view(1, 1, 5, 5)
wp = warp_perspective(x, I, (1, 4), align_corners=True)
print("warp_perspective(x, I, (1,4), align_corners=True) =", wp.flatten().tolist())
cb = crop_by_transform_mat(x, I, (1, 4), align_corners=True)
print("crop_by_transform_mat(x, I, (1,4), align_corners=True) =", cb.flatten().tolist())
```

## Observed

```
# 1. the helper itself
normal_transform_pixel(4, 1) =
 tensor([[[ 2.0000e+14,  0.0000e+00, -1.0000e+00],
         [ 0.0000e+00,  6.6667e-01, -1.0000e+00],
         [ 0.0000e+00,  0.0000e+00,  1.0000e+00]]])
normal_transform_pixel(1, 1) flat = [200000000753664.0, 0.0, -1.0, 0.0, 200000000753664.0, -1.0, 0.0, 0.0, 1.0]
  h=4,w=0 -> sx=-2.0, sy=0.6666666865348816, finite=True
  h=0,w=5 -> sx=0.5, sy=-2.0, finite=True
  h=0,w=0 -> sx=-2.0, sy=-2.0, finite=True
  h=4,w=-3 -> sx=-0.5, sy=0.6666666865348816, finite=True

# 2. eps guards only the ==1 branch, not the ==0 one
  eps=1.0,  W==1 -> sx = 2.0
  eps=1.0,  W==0 -> sx = -2.0
  eps=1e-14,W==5 -> sx = 0.5

# 3. it propagates into the public homography API
normalize_homography(I,(4,1),(4,5))   = [2.500000167887412e-15, 0.0, -1.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0]
denormalize_homography(I,(4,1),(4,5)) = [400000001507328.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0]
normalize_homography(I,(4,0),(4,5))   = [-0.25, 0.0, -1.25, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0]

# 4. and into 3D, where depth==1 gives a numerically non-invertible matrix
normalize_homography3d(I,(1,4,5),(3,8,9)) row2 = [0.0, 0.0, 4.99999991225835e-15, -1.0]
  o[2,2] = 4.99999991225835e-15  det = 1.0714285980035544e-15
  torch.linalg.inv succeeds: True
normal_transform_pixel3d(1,4,5) diag = [0.5, 0.6666666865348816, 200000000753664.0]

# 5. the user-visible end of the chain: NaN out of warp_perspective (align_corners=TRUE)
warp_perspective(x, I, (1,4), align_corners=True) = [nan, nan, nan, nan]
crop_by_transform_mat(x, I, (1,4), align_corners=True) = [nan, nan, nan, nan]
```

Note the 3D case in block 4: `depth == 1` produces a z scale of `4.99999991e-15` — the reciprocal of the `2e14` blow-up, and **not** an exact zero (it merely renders as `0.` at printed precision). `det = 1.07e-15`, so `torch.linalg.inv` still succeeds: the matrix is non-invertible *in practice*, not formally singular, and it annihilates every z coordinate while looking well-formed.

## Expected

A degenerate size should either raise or produce a defined, finite, non-destructive transform. Concretely:

- **`size == 1`.** There is a mathematically correct answer here that the `eps` hack is standing in for: a one-pixel axis has no extent, so the only sensible normalized coordinate for its single pixel is the centre of the range. Setting `sx = 0`, offset `0` (i.e. mapping the lone pixel to `0.0`, the centre of the normalized range) is finite and keeps the transform well-defined. Returning `2e14` instead is what turns into the NaN in block 5. The exact choice is a maintainer call — the requirement is only that it be finite and non-destructive.
- **`size == 0`.** An empty axis is not a coordinate system; `−2.0` (a mirror) is not a defensible answer under any convention. This should raise, matching the direction taken for the sibling functions in #3940.
- **negative sizes.** Should raise. A negative image dimension is a caller bug, and today it silently yields a mirrored normalization.
- **`eps`.** Whatever is decided above, the docstring's `epsilon to prevent divide-by-zero errors` should describe what the parameter does. Today it is consulted only when `size == 1`, i.e. never in the case where the denominator is actually at risk of being zero.

The same three fixes apply verbatim to `normal_transform_pixel3d`.

## Relationship to existing issues

- **#3929** is the *symptom* of this root in one caller: `crop_by_transform_mat` with a 1-pixel output size returns all-NaN. That issue does not name `normal_transform_pixel` and proposes no fix there. Block 5 above is the executed chain that connects them — `normal_transform_pixel(1, 4)` → `normalize_homography(..., (1, 4))` → `warp_perspective(..., (1, 4))` all-NaN → `crop_by_transform_mat(..., (1, 4))` all-NaN — so this issue supplies #3929's root cause and should be fixed with it in view rather than duplicating it. **Cross-referencing #3929, not superseding it.**
- **#3940** is the *same defect class* in `normalize_pixel_coordinates` / `denormalize_pixel_coordinates` — a different function pair with its own copy of the degenerate-denominator logic. Sibling issue, and whatever policy is chosen there (raise vs finite fallback) should be chosen here too so the two families stay consistent.
- **#3904** — see the scope note at the top. Different defect, different branch, reserved.
- #3937 / #3938 / #3939 are unrelated.

## Classification

**Broken output**, in two of the three cases unambiguously: `2e14` and the downstream all-NaN warp are wrong under every convention and every `align_corners` setting, and per the reference-core lifecycle (Part I / Part III §2) NaN-producing paths may land as ordinary bugfixes without waiting for the coordinated Repair & Deprecation window.

The `size == 0` / negative-size half is a **semantic change** by comparison — those calls currently return silently, so making them raise turns working-but-meaningless calls into errors. That half should follow whatever #3940 settles on. Every non-degenerate size is unaffected either way, so the repair can be landed under #3916-style byte-identity discipline on the normal path.

## Environment

kornia 0.9.0rc1 @ `3358ef16`, torch 2.9.1, cpu. Executed from a worktree carrying batch-3a docs/tests commits on top of `3358ef16`; `normal_transform_pixel`, `normal_transform_pixel3d`, `normalize_homography`, `denormalize_homography` and `normalize_homography3d` are all byte-identical to `origin/main` there (no diff hunk lands in any of them; function bodies compare AST-identical).

Posted on behalf of @ducha-aiki by Claude (Fable 5).
